### ISSUE TITLE
extern Swift declaration not public

### ISSUE THREAD

**Issue by tmolitor-stud-tu** [chinedufn/swift-bridge#241]
We are building a library that has swift code that should be called by rust in an extra swift file (besides the auto generated ones).
But we keep getting linker errors that `__swift_bridge__$rust_handler` could not be found.
Until we make the auto-generated function public:
```
@_cdecl("__swift_bridge__$rust_handler")
public func __swift_bridge__rust_handler (_ text: UnsafeMutableRawPointer) {
    rust_handler(text: RustString(ptr: text))
}
```
Could you fix the code autogenerator to make these public by default (or at least introduce some compile-time switch to do that)?

**Comment by NiwakaDev** [comment]
Could you share the minimal example?

**Comment by tmolitor-stud-tu** [comment]
See [this zip file](https://github.com/chinedufn/swift-bridge/files/12553324/example.zip) for a minimal example.
The build-rust.sh script is based upon https://github.com/chinedufn/swift-bridge/tree/master/book/src/building/swift-packages

The generated code looks like this:
```
import RustXcframework
public func test() {
    __swift_bridge__$test()
}
@_cdecl("__swift_bridge__$swift_handler")
func __swift_bridge__swift_handler (_ text: UnsafeMutableRawPointer) {
    swift_handler(text: RustString(ptr: text))
}
```
Thus the linker will report `__swift_bridge__$swift_handler` as missing once the generated `LibTestBridge` package is linked to from any swift project.
Adding `public` in front of the `__swift_bridge__swift_handler` definition will solve that issue.

That's because of swifts [access control](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/accesscontrol/) defaulting to `internal` which prevents the binary `RustXcframework.xcframework` from accessing the `__swift_bridge__swift_handler` supposed to be exposed as symbol `__swift_bridge__$swift_handler` in the object file generated by the swift compiler, **but the swift compiler does not expose this symbol**.

The swift compiler does not expose the symbol because the `internal` access control means this function should only be accessible by the swift module it resides in, not by any other modules (linked in).

**Comment by tmolitor-stud-tu** [comment]
If you point me in the right direction, I can create a pull request fixing this :)

**Comment by NiwakaDev** [comment]
@tmolitor-stud-tu 
Sorry for the response and thank you for sharing the example.

> If you point me in the right direction, I can create a pull request fixing this :)

 I'll dive into this issue next weekend because I'm a bit busy right now.

**Comment by NiwakaDev** [comment]
@tmolitor-stud-tu 

```rust
public func test() {
    __swift_bridge__$test()
}
@_cdecl("__swift_bridge__$swift_handler")
func __swift_bridge__swift_handler (_ text: UnsafeMutableRawPointer) {
    swift_handler(text: RustString(ptr: text))
}
```

In your example, it seems like `__swift_bridge__swift_handler` isn't called in other swift modules. Do you want to call this method in other swift modules?

**Comment by NiwakaDev** [comment]
As far as I remember, this method is used on the Rust side. So, other swift modules shouldn't call this.

**Comment by tmolitor-stud-tu** [comment]
> As far as I remember, this method is used on the Rust side. So, other swift modules shouldn't call this.

Exactly, but the binary rust XCFramework has an unresolved symbol (`__swift_bridge__$swift_handler`) because the `__swift_bridge__swift_handler` swift function generated by swift-bridge is not public and thus its symbol won't be exported.
This means the linker sees two binaries, one containing the rust stuff that wants a symbol named `__swift_bridge__$swift_handler` and the other (the swift one) not exporting that symbol even though it should (that's the whole point of the `__swift_bridge__swift_handler` function in the first place: being there to be called by rust code).

Please note: even though I did not include swift code containing the `swift_handler` function, that linker problem is unrelated to this. Just create a swift file inside the `test-bridge/generated` directory containing that function.

Sidenote: a full (non-minimal) example can be found over here: https://github.com/monal-im/Monal/tree/develop/rust
Please note the `sed` invocation at the end of the build script fixing this issue: https://github.com/monal-im/Monal/blob/develop/rust/build-rust.sh#L60

**Comment by NiwakaDev** [comment]
@tmolitor-stud-tu 


What linker do you use?

> a full (non-minimal) example can be found over here: https://github.com/monal-im/Monal/tree/develop/rust
Please note the sed invocation at the end of the build script fixing this issue: https://github.com/monal-im/Monal/blob/develop/rust/build-rust.sh#L60

By the way, I added the your package to my minimal project like this:
<img width="884" alt="image1" src="https://github.com/chinedufn/swift-bridge/assets/61189782/2eab8157-e997-4a9c-b30a-e7cbcb692d75">
<img width="544" alt="image2" src="https://github.com/chinedufn/swift-bridge/assets/61189782/99c17647-331d-47a7-8ad3-c9fb4e5dd323">

Even though I don't put `public` in front of `__swift_bridge__rust_panic_handler` method, Xcode`s default linker could link them. Am I missing something?

## Environment


Xcode: Version 14.3.1

**Comment by NiwakaDev** [comment]
> This means the linker sees two binaries, one containing the rust stuff that wants a symbol named __swift_bridge__$swift_handler and the other (the swift one) not exporting that symbol even though it should (that's the whole point of the __swift_bridge__swift_handler function in the first place: being there to be called by rust code).

As you can see the immediately above comment, it seems like the programs run, which means the linker could link rust object files and swift object files.

**Comment by chinedufn** [comment]
https://github.com/chinedufn/swift-bridge/pull/262 should fix this

**Comment by tmolitor-stud-tu** [comment]
Thanks @jmp-0x7C0 !!!

**Comment by JohnnyTarrr** [comment]
Following up on this issue — we're hitting the same symptom on **swift-bridge 0.1.59** (latest published, 2026-01-06), well after PR #262 merged. Filing fresh evidence in case PR #262's fix was incomplete or applies to only some code paths.

## Reproducer (2026-05-13)

Crate uses `extern "Swift"` functions with `String -> String` and primitive signatures. Build via `cargo build --release --features swift-runtime`, then `swiftc -emit-library` to produce a macOS dylib.

## Generated Swift wrapper (0.1.59 output)

```swift
@_cdecl("__swift_bridge__$cryptokit_sha256_hex")
func __swift_bridge__cryptokit_sha256_hex (_ hex_input: UnsafeMutableRawPointer) -> UnsafeMutableRawPointer { ... }
```

Note: no `public` modifier. This is the same shape that #262 was thought to address.

## Symptom (still present on 0.1.59)

```
$ nm libLydoCryptoKit.dylib | grep cryptokit_sha256_hex
0000000000004f10 t ___swift_bridge__$cryptokit_sha256_hex
```

Lowercase `t` = local symbol. The Rust test binary then fails to link:

```
$ cargo test --release --features swift-runtime
  Undefined symbols for architecture arm64:
    "___swift_bridge__$cryptokit_sha256_hex", referenced from: ...
```

## Workaround (current)

`sed` patch in our `scripts/host-smoke.sh` post-codegen:

```bash
sed -i '' 's/^func __swift_bridge__/public func __swift_bridge__/' \
    generated/lydo-mls-cryptokit/lydo-mls-cryptokit.swift
```

Idempotent, adds ~50ms to the host smoke. Works around it indefinitely, but the upstream generator should emit `public func` directly.

## Environment

- swift-bridge **0.1.59**
- swift 6.2.3 (Apple Swift)
- macOS 26.0 / aarch64-apple-darwin
- Build target: standalone macOS dylib (`swiftc -emit-library`), then linked from `cargo test` binary

## Hypothesis

PR #262 may have fixed this for the xcframework code path (where Swift module + Rust consumer compile together inside the same target) but not the standalone-dylib code path (where Swift dylib is built first, then linked from a separate Rust consumer). The two compile modes have different symbol-visibility behavior — internal symbols stay visible inside a single module but get stripped from dylib exports.

If a maintainer wants a minimal reproducer (`cargo init` + 5-line `bridge.rs` + the `swiftc` command + `nm` check), happy to put one together. Just wanted to surface that the issue persists in 0.1.59 in case it's useful to reopen.

Thanks for the great library!

**Comment by chinedufn** [comment]
PR #262 was never merged
